pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,399 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package compactor
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"path"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gorilla/mux"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+
+	"github.com/grafana/dskit/tenant"
+	"github.com/grafana/regexp"
+
+	"github.com/grafana/mimir/pkg/storage/bucket"
+	mimir_tsdb "github.com/grafana/mimir/pkg/storage/tsdb"
+	"github.com/grafana/mimir/pkg/util"
+	util_log "github.com/grafana/mimir/pkg/util/log"
+)
+
+const tmpMetaFilename = "uploading-" + block.MetaFilename
+
+var rePath = regexp.MustCompile(`^(index|chunks/\d{6})$`)
+
+// HandleBlockUpload handles requests for starting or completing block uploads.
+//
+// The query parameter uploadComplete (true or false, default false) controls whether the
+// upload should be completed or not.
+//
+// Starting the uploading of a block means to upload meta file and verify that the upload can go ahead.
+// In practice this means to check that the (complete) block isn't already in block storage,
+// and that meta file is valid.
+func (c *MultitenantCompactor) HandleBlockUpload(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	blockID := vars["block"]
+	if blockID == "" {
+		http.Error(w, "missing block ID", http.StatusBadRequest)
+		return
+	}
+	bULID, err := ulid.Parse(blockID)
+	if err != nil {
+		http.Error(w, "invalid block ID", http.StatusBadRequest)
+		return
+	}
+	tenantID, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+	if err != nil {
+		http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+		return
+	}
+
+	logger := util_log.WithContext(ctx, c.logger)
+	logger = log.With(logger, "block", blockID)
+
+	userBkt := bucket.NewUserBucketClient(tenantID, c.bucketClient, c.cfgProvider)
+	if err := checkForComplete(ctx, logger, bULID, userBkt); err != nil {
+		var httpErr httpError
+		if errors.As(err, &httpErr) {
+			level.Warn(logger).Log("msg", httpErr.message, "err", err)
+			http.Error(w, httpErr.message, httpErr.statusCode)
+			return
+		}
+
+		level.Error(logger).Log("msg", "an unexpected error occurred", "err", err)
```

Comment:
This message is very generic. Error will include "failed to check existence of meta.json in object storage", but it still gives no info on what operation was done (file upload? block finish?), and same message is used elsewhere in the code. Can eg. add "operation" key to the logger, or expand the message a bit?